### PR TITLE
Introduce Behavior#unsafeCast (asInstanceOf alt)

### PR DESCRIPTION
**Pull request by dwijnand**
Also:
* Introduce StoppedBehavior#cast, as a local, safe alternative.
* Remove some unnecessary casts in WithMdcBehaviorInterceptor
* Reuse Behavior.same & Behavior.unhandled

### CODE REVIEW

File: akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala [hunk 1]
```diff
@@ -44,6 +44,16 @@ sealed abstract class Behavior[T] { behavior ⇒
...
 
+  /**
+   * Unsafe utility method for changing the type accepted by this Behavior;
+   * provided as an alternative to the universally available `asInstanceOf`, which
+   * casts the entire type rather than just the type parameter.
+   * Typically used to upcast a type, for instance from `Nothing` to some type `U`.
+   * Use it with caution, it may lead to a [[ClassCastException]] when you send a message
+   * to the resulting [[Behavior[U]]].
+   */
+  final def unsafeCast[U]: Behavior[U] = this.asInstanceOf[Behavior[U]]
```
Comment (hepin1989) [inline akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala]: great！

Comment (hepin1989) [reply akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala]: I introduced this once in https://github.com/akka/akka/pull/26024

**Review by chbatey — APPROVED** [review]
LGTM, thanks

File: akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala (hunk 1 above)
Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala]: Should it be `unsafeUpcast` in the same way as `ActorRef.unsafeUpcast`. `narrow` should be used for the other direction. Are we sure this should be exposed to users, or should we start out with making it internal API?

File: akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala [hunk 2]
```diff
@@ -252,6 +262,8 @@ object Behavior {
   private[akka] class StoppedBehavior[T](val postStop: OptionVal[Behavior[T]]) extends Behavior[T] {
     validatePostStop(postStop)
 
+    def cast[U]: Behavior[U] = unsafeCast[U] // won't receive more messages so cast is safe
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala]: why is this needed? why not use unsafeCast directly?

Comment (dwijnand) [reply akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala]: Not strictly needed.  Can be just unsafeCast with the justifying comment, instead.

File: akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala (hunk 1 above)
Comment (dwijnand) [reply akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala]: Not every cast is to a super-type, which is why it isn't `unsafeUpcast`.

I'm happy with making it internal API, which formulation do you prefer?  `@InternalApi`? `private[akka]`? Separate implicit class? Separate function?

Comment (patriknw) [reply akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala]: ok

```
  /**
   * INTERNAL API
   */
  @InternalApi private[akka]
```

**Review by patriknw — APPROVED** [review]
LGTM
